keps/sig-apps/4017-pod-index-label/README.md

```diff
@@ -0,0 +1,891 @@
+analogous to the pod completion index for Indexed Jobs).
+
+Indexed Jobs set the pod index as an annotation, which means it can be utilized by the downward API,
+however, it cannot easily be used for selecting pods based on their index or filtering metrics/logs, 
+or for doing things like routing network traffic to a pod with a specific index (e.g., index 0). 
+
+StatefulSets include the pod index as part of the pod name. This means that the only way for a 
+StatefulSet pod to know its own index is to [parse the pod name](https://github.com/kubernetes/kubernetes/blob/b31774d39b8b493a9468c030e598d8520308443d/pkg/controller/statefulset/stateful_set_utils.go#L63-L75),
+which is somewhat hacky solution. Furthermore, the pod index cannot be utilized by the downward API,
+nor used for use cases such as filtering metrics/logs or routing traffic to a pod with a specific index.
+
+To address these issues, we propose to set the pod index as a **label** on pods for both StatefulSets and
+Indexed Jobs. This will allow the pod index to be utilized via the downward API, so StatefulSet pods
+can easily know their own index. In addition, this will allow metrics and logs to more easily be
+filtered by pod index, for both StatefulSets and Indexed Jobs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Set pod index as label on pods for StatefulSet/Indexed Jobs.
+- Adding the label should not be disruptive to existing workloads.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+At a high level, the proposal is to modify the StatefulSet and Job controllers to set the pod index
+as a pod label at pod creation time (for jobs, this would only apply to jobs in
+Indexed completion mode). The details of this are outlined in the Design Details section below.
+
+StatefulSet pod label: `statefulset.kubernetes.io/pod-index`
+Indexed Job pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+
+#### Story 1
+As a user, I would like to lookup a job's pod logs by its index.
+
+#### Story 2
+As a user, I would like to target traffic to a specific pod index (e.g., index 0) in a StatefulSet.
```

Comment:
Also for Indexed Job. In particular, for architectures where all workers connect to a single head pod.

Instead of creating a Service that matches the entire Job, you can create a Service that matches only the head Pod, which would be more performant, specially with a high number of workers. Performance should be part of the motivation.

Also, please make sure to update this doc https://kubernetes.io/docs/tasks/job/job-with-pod-to-pod-communication/ to list this pattern

Reply:
Good point, updated story 2 to include this.